### PR TITLE
Add Parameter wrapper to support older Android versions

### LINKED ISSUE #2: Support for Android pre-API 26
I'm trying to migrate our codebases from Burst in an effort to deprecate Burst, but this trips us up pretty hard on Android.

I found this comment:

https://github.com/google/TestParameterInjector/blob/7a62c43af8c4f10f1451ec84cd21c2ba509dae82/src/main/java/com/google/testing/junit/testparameterinjector/TestParameterAnnotationMethodProcessor.java#L902-L905

We could _probably_ get this method and associated interface into D8/R8 desugaring since most (all?) of `Parameter` can be emulated by wrapping the originating `Executable` and delegating to its methods. Have you spoken to the R8 team at all about this?

Or would you be open to taking that approach for now while separately we try and convince the R8 team? Since `AnnotationWithMetadata` is already an abstraction the only change would seem to be to duplicate the `getAnnotationWithMetadataListWithType` and have it accept the `Executable` subtypes (`Constructor`, `Method`, etc.) to pull out the needed components.

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #2.

**Note**: I haven't tested this on an actual pre-26 Android device yet. Will do later tonight or tomorrow. Just wanted to get the conversation started on the workaround.

### CODE REVIEW

File: src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      return Legacy.create(constructor);
+    }
+  }
+
+  @SuppressWarnings("AndroidJdkLibsChecker") // j.l.r.Parameter is not available on old Android SDKs.
+  static ParameterWrapper[] get(Method method) {
+    try {
+      return Java8.create(method.getParameters());
+    } catch (NoSuchMethodError ignored) {
+      return Legacy.create(method);
```
Comment (nymanjens) [inline src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java:38]: Why not always use the Legacy version?

Comment (JakeWharton) [reply src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java:38]: The `Java8` wrapper correctly propagates parameter names (assuming they're available in the class file). I did not look to see how the name is used in practice, though.

**Comment by nymanjens** [comment]
Thanks a lot for this PR!

So this PR avoids a dependency on `java.lang.reflect.Parameter`, which these older Android versions don't support.

FYI: [TestParametersMethodProcessor.java](https://github.com/google/TestParameterInjector/blob/main/src/main/java/com/google/testing/junit/testparameterinjector/TestParametersMethodProcessor.java) also has a dependency on `java.lang.reflect.Parameter`:
* `TestParametersMethodProcessor` is used for `@TestParameters`.
* `TestParameterAnnotationMethodProcessor` edited in this PR is used for `@TestParameter`.

However, `@TestParameters` can only work if the parameter names are stored into the jar by the compiler, which normally happens by enabling a command line flag in the Java compiler. I suspect this might in itself be a tricky requirement for the older Android versions?

In any case, I think this runner can be quite useful, even without `@TestParameters`.

File: src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java (hunk 1 above)
Comment (nymanjens) [reply src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java:38]: The parameter name is to (optionally) format the name of the test:

```
myTest[myParameter=true]
```

In any case, I've been investigating whether we can use `AnnotationWithMetadata` a bit earlier in the process. I'm fine with merging this PR for now and see if I can improve on it at a later point.

**Review by nymanjens — COMMENTED** [review]
I asked a colleague who added Anroid support to have a look. Since he's from the US, I'll allow him to get back to me first.

**Comment by JakeWharton** [comment]
> I suspect this might in itself be a tricky requirement for the older Android versions?

Even if you opt-in to parameter name retention, Android's toolchain drops the information because there's nowhere to put it in the Dalvik bytecode format. So yeah it sounds like that functionality is going to be a complete non-starter.

**Review by nymanjens — APPROVED** [review]

**Comment by graememorgan** [comment]
LGTM, thanks!

**Pull request merged**